### PR TITLE
Implemented Communication job

### PR DESCRIPTION
**Pull request by leobcgov**
## Status
**Ready**

## Fixes or Implements [AB#8638](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/8638)
- [X] Enhancement
- [ ] Bug
- [ ] Documentation

## Description:
- Added Communication Job which is scheduled for every 30 minutes to create Emails which need to be sent to the Gateway's active users

## Testing
- [ ] Unit Tests Updated
- [ ] Functional Tests Updated
- [X] Not Required

### Steps to Reproduce:
If this is a bug, please provide details on how to reproduce the code.

### UI Changes
NO

### Browsers Tested
- [ ] Chrome
- [ ] Safari
- [ ] Edge
- [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments
- Testing performed locally.
- Migration Script tested:
image

### CODE REVIEW

File: Apps/Database/src/Constants/EmailConsts.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Represents the To of emails.
+    /// </summary>
+    public static class EmailConsts
+    {
+        /// <summary>
+        /// Constant value to represent undisclosed-recipients of the To field of the email.
+        /// </summary>
+        public const string FromEmailAddressHGDonotreply = "HG_Donotreply@gov.bc.ca";
```
Comment (sslaws) [inline Apps/Database/src/Constants/EmailConsts.cs:26]: This should be configuration vs a constant as it may change in the future.

File: Apps/Database/src/Models/Communication.cs [hunk 2]
```diff
@@ -74,5 +75,18 @@ public class Communication : AuditableEntity
         /// The lower the value the lower the priority.
         /// </summary>
         public int Priority { get; set; } = EmailPriority.Standard;
+
+
+#pragma warning disable CA2227 // Collection properties should be read only
```
Comment (sslaws) [inline Apps/Database/src/Models/Communication.cs:80]: We should be fixing these issues instead of using a pragma to disable them

File: Apps/Database/src/Models/Email.cs [hunk 3]
```diff
@@ -99,5 +100,17 @@ public class Email : AuditableEntity
         [Required]
         [MaxLength(10)]
         public string EmailStatusCode { get; set; } = EmailStatus.New;
+
+#pragma warning disable CA2227 // Collection properties should be read only
```
Comment (sslaws) [inline Apps/Database/src/Models/Email.cs:104]: ditto on the pragmas from earlier

File: Apps/Database/src/Models/UserProfile.cs [hunk 4]
```diff
@@ -72,5 +73,17 @@ public class UserProfile : AuditableEntity
         /// </summary>
         [MaxLength(44)]
         public string? EncryptionKey { get; set; }
+
+#pragma warning disable CA2227 // Collection properties should be read only
```
Comment (sslaws) [inline Apps/Database/src/Models/UserProfile.cs:77]: and again

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+                hdidsWithMostRecentCreatedCommEmails = hdids.ToString();
+            }
+
+            var userProfiles = this.dbContext.UserProfile
+                .Where(profile => !profile.ClosedDateTime.HasValue
+                        && !string.IsNullOrEmpty(profile.Email)
+                        && (!createdOnOrAfter.HasValue || profile.CreatedDateTime >= createdOnOrAfter)
+
+#pragma warning disable CA1307 // Specify StringComparison
```
Comment (sslaws) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:105]: NIT:  I like using annotations to disable over pragmas as they look cleaner in the code.

**Comment by leobcgov** [comment]
@sslaws my latest check-in contains the fixes re. your comments, can you please review them? thanks!

File: Apps/Database/src/Models/CommunicationEmail.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        /// </summary>
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        [Column("CommunicationEmailId")]
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Communication Id (foreign key).
+        /// </summary>
+        [ForeignKey("Communication")]
```
Comment (sslaws) [inline Apps/Database/src/Models/CommunicationEmail.cs:38]: Kudo's as I don't think anyone has done this before and it saves custom code in the context.

File: Apps/Database/src/Models/UserProfile.cs [hunk 7]
```diff
@@ -72,5 +73,10 @@ public class UserProfile : AuditableEntity
         /// </summary>
         [MaxLength(44)]
         public string? EncryptionKey { get; set; }
+
+        /// <summary>
+        /// Gets the list containing all the communication emails related to this user profile.
+        /// </summary>
+        public virtual List<CommunicationEmail> CommunicationEmails { get; } = null!;
```
Comment (sslaws) [inline Apps/Database/src/Models/UserProfile.cs:80]: We did this a lot early on but we've been leaning against it lately and instead have been creating empty lists.  We should discuss during grooming.

Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs:80]: Oh I didn't know that. One benefit of this is ability to include the list or a filtered list of child records in the LINQ query (i.e. .Include(...)

Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs:80]: updated to "public virtual List<CommunicationEmail> CommunicationEmails { get; } = new List<CommunicationEmail>();"

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                StringBuilder hdids = new StringBuilder();
+                foreach (var commEmail in commEmails)
+                {
+                    hdids.Append(commEmail.UserProfileHdId + "|");
+                }
+
+                hdidsWithMostRecentCreatedCommEmails = hdids.ToString();
+            }
+
+            var userProfiles = this.dbContext.UserProfile
```
Comment (sslaws) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:100]: I think we could likely optimize this query. Can you create a task in azure for the future?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                        {
+                            if (usersToSendCommEmails != null && usersToSendCommEmails.Count > 0)
+                            {
+                                lastProcessedProfileHdid = usersToSendCommEmails[usersToSendCommEmails.Count - 1].HdId;
+                            }
+
+                            usersToSendCommEmails = this.commEmailDelegate.GetActiveUserProfilesWithoutCommEmailByCommunicationIdAndByCreatedOnOrAfter(comm.Id, createdOnOrAfterFilter, this.retryFetchSize);
+                            foreach (UserProfile profile in usersToSendCommEmails)
+                            {
+                                // Idea for Improvement: we can look into a lib for Bulk Insert / Update using .NET Core for Postgres (if needed).
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:99]: Yes, bulk insert operation would be faster.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                                this.commEmailDelegate.Add(commEmail, false);
+                            }
+
+                            if (usersToSendCommEmails.Count > 0)
+                            {
+                                // Set the createdOnOrAfterFilter for the next query retrieving the next chunk of user profiles.
+                                createdOnOrAfterFilter = usersToSendCommEmails[usersToSendCommEmails.Count - 1].CreatedDateTime;
+                            }
+
+                            this.dbContext.SaveChanges(); // commit after every retryFetchSize (or 250) pairs of Email & CommunicationEmail.
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:128]: Now that I see this here, I see that we’re missing a state... we should flip the state to processing or something similar when we start and the outer query should look for new and processing’s comms.

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs (hunk 8 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:100]: Created 8717 for this improvement.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 10 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:128]: I will add a new "Processing" status to the Communication's Communication Status Code and update it when processing a communication in the job scheduler.

Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:133]: Code updated per @sslaws 's great comments above.

File: Apps/Database/src/Context/GatewayDbContext.cs [hunk 11]
```diff
@@ -687,6 +688,15 @@ private void SeedCommunication(ModelBuilder modelBuilder)
                     UpdatedDateTime = this.DefaultSeedDate,
                 },
                 new CommunicationStatusCode
+                {
+                    StatusCode = CommunicationStatus.Processing,
+                    Description = "A Communication is in process by batch",
```
Comment (FuriousLlama) [inline Apps/Database/src/Context/GatewayDbContext.cs:693]: nit Communication is being processed

File: Apps/Database/src/Delegates/DBCommunicationDelegate.cs [hunk 12]
```diff
@@ -132,5 +133,15 @@ public DBResult<Communication> Update(Communication communication, bool commit =
             this.logger.LogDebug($"Finished updating Communication in DB");
             return result;
         }
+
+        /// <inheritdoc />
+        public List<Communication> GetEmailCommunicationsInNewProcessingOrError()
+        {
+            this.logger.LogTrace($"Getting Communications by Type and Status Code from DB...");
+            List<Communication> retVal = this.dbContext.Communication.Where(c => c.CommunicationTypeCode == CommunicationType.Email && (c.CommunicationStatusCode == CommunicationStatus.New || c.CommunicationStatusCode == CommunicationStatus.Processing || c.CommunicationStatusCode == CommunicationStatus.Error))
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/DBCommunicationDelegate.cs:141]: Instead, we can search for not processed.

File: Apps/Database/src/Delegates/DBCommunicationDelegate.cs [hunk 13]
```diff
@@ -132,5 +133,15 @@ public DBResult<Communication> Update(Communication communication, bool commit =
             this.logger.LogDebug($"Finished updating Communication in DB");
             return result;
         }
+
+        /// <inheritdoc />
+        public List<Communication> GetEmailCommunicationsInNewProcessingOrError()
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/DBCommunicationDelegate.cs:138]: should we not get only the communications for today or past its scheduled date?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        /// <inheritdoc />
+        [DisableConcurrentExecution(ConcurrencyTimeout)]
+        public void CreateCommunicationEmailsForNewCommunications()
+        {
+            this.logger.LogDebug($"Creating emails & communication emails...");
+            List<Communication> communications = this.communicationDelegate.GetEmailCommunicationsInNewProcessingOrError();
+
+            if (communications.Count > 0)
+            {
+                this.logger.LogInformation($"Found {communications.Count} communications which are in New status.");
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:79]: Not necessarily true, the query gets other statuses as well

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+                                    UserProfile = profile,
+                                    Communication = comm,
+                                };
+                                this.commEmailDelegate.Add(commEmail, false);
+                            }
+
+                            if (usersToSendCommEmails.Count > 0)
+                            {
+                                // Set the createdOnOrAfterFilter for the next query retrieving the next chunk of user profiles.
+                                createdOnOrAfterFilter = usersToSendCommEmails[usersToSendCommEmails.Count - 1].CreatedDateTime;
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:124]: why is this necessary? Does the order of the profiles matter?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+            Contract.Requires((configuration != null) && (emailQueueService != null));
+            this.configuration = configuration!;
+            this.logger = logger;
+            this.communicationDelegate = communicationDelegate!;
+            this.commEmailDelegate = commEmailDelegate!;
+            this.emailQueueService = emailQueueService!;
+            this.dbContext = dbContext;
+
+            IConfigurationSection emailJobSection = this.configuration.GetSection("EmailJob");
+            this.retryFetchSize = emailJobSection.GetValue<int>("MaxRetryFetchSize", 250);
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:65]: retryFetchSize is confusion. How about "ProcessMaxBatchSize"

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+                .Where(profile => !profile.ClosedDateTime.HasValue
+                        && !string.IsNullOrEmpty(profile.Email)
+                        && (!createdOnOrAfter.HasValue || profile.CreatedDateTime >= createdOnOrAfter)
+                        && hdidsWithMostRecentCreatedCommEmails != profile.HdId)
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:104]: Not sure we need the hdidsWithMostRecentCreatedCommEmails parameter. Instead, the query can select for profiles that do not have a communication entry

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+            var commEmails = this.dbContext.CommunicationEmail.Where(c => c.CommunicationId == communicationId).OrderByDescending(c => c.CreatedDateTime).Take(maxRows)
+                    .Include(commEmail => commEmail.UserProfile).Where(c => !createdOnOrAfter.HasValue || c.UserProfile.CreatedDateTime >= createdOnOrAfter).ToList();
+            string hdidsWithMostRecentCreatedCommEmails = string.Empty;
+            if (commEmails.Count > 0)
+            {
+                createdOnOrAfter = commEmails[0].UserProfile.CreatedDateTime;
+                StringBuilder hdids = new StringBuilder();
+                foreach (var commEmail in commEmails)
+                {
+                    hdids.Append(commEmail.UserProfileHdId + "|");
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:94]: why is the concatenation necessary?

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+                    result.Message = e.Message;
+                }
+            }
+
+            this.logger.LogDebug($"Finished adding Communication in DB");
+            return result;
+        }
+
+        /// <inheritdoc />
+        public List<UserProfile> GetActiveUserProfilesWithoutCommEmailByCommunicationIdAndByCreatedOnOrAfter(Guid communicationId, DateTime? createdOnOrAfter = null, int maxRows = 500)
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:80]: kinda nit: 76 chars = Overly long method name?

File: Apps/Database/src/Delegates/ICommunicationEmailDelegate.cs [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+        /// <summary>
+        /// Get a list of active user profiles which are not linked to any communication email by the specified Communication Id
+        ///     AND were created on of after a specific date and time .
+        /// </summary>
+        /// <param name="communicationId">The communication id for filtering communication emails.</param>
+        /// <param name="createdOnOrAfter">The profiles must have created on or after to this date.</param>
+        /// <param name="maxRows">The maximum amount of user profiles to return.</param>
+        /// <returns>A list of communication emails of the specified Communication, wrapped in a DBResult.</returns>
+        List<UserProfile> GetActiveUserProfilesWithoutCommEmailByCommunicationIdAndByCreatedOnOrAfter(Guid communicationId, DateTime? createdOnOrAfter = null, int maxRows = 500);
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/ICommunicationEmailDelegate.cs:44]: This looks like an anti-pattern. The method is only being used in one place but has two optional parameters

File: Apps/Database/src/Models/Communication.cs [hunk 21]
```diff
@@ -74,5 +75,10 @@ public class Communication : AuditableEntity
         /// The lower the value the lower the priority.
         /// </summary>
         public int Priority { get; set; } = EmailPriority.Standard;
+
+        /// <summary>
+        /// Gets the list containing all the communication emails related to this communication.
+        /// </summary>
+        public virtual List<CommunicationEmail> CommunicationEmails { get; } = null!;
```
Comment (FuriousLlama) [inline Apps/Database/src/Models/Communication.cs:82]: ```= new List<CommunicationEmail>;```

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+                                    Body = comm.Text,
+                                    FormatCode = EmailFormat.HTML,
+                                    Priority = comm.Priority,
+                                };
+                                this.emailQueueService.QueueNewEmail(email, false);
+
+                                // Inser a new CommunicationEmail record into
+                                CommunicationEmail commEmail = new CommunicationEmail()
+                                {
+                                    Email = email,
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: This email object does not have an ID. Does that work when adding it to the table on line 118?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+        public void CreateCommunicationEmailsForNewCommunications()
+        {
+            this.logger.LogDebug($"Creating emails & communication emails...");
+            List<Communication> communications = this.communicationDelegate.GetEmailCommunicationsInNewProcessingOrError();
+
+            if (communications.Count > 0)
+            {
+                this.logger.LogInformation($"Found {communications.Count} communications which are in New status.");
+
+                foreach (Communication comm in communications)
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:81]: nit: comm => communication

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 16 above)
Comment (sslaws) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:65]: It is really just fetch size, we're pulling a block to process and the majority of the time the block should be processed without retrying.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 22 above)
Comment (sslaws) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: Yes that works, the DB creates the ID and EF gets the ID when it inserts it and then uses it for inserts down the line.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 15 above)
Comment (sslaws) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:124]: The order is being used for processing blocks as the job will read from the last date read.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+                        // Update Communication Status to Processed
+                        comm.CommunicationStatusCode = CommunicationStatus.Processed;
+                        this.communicationDelegate.Update(comm, true);
+                    }
+                    catch (Exception e)
+                    {
+                        // log the exception as a warning but we can continue
+                        this.logger.LogWarning($"Error while creating new emails and communication email records for Email Communication - skipping for now\n{e.ToString()}");
+
+                        if (comm.CommunicationStatusCode != CommunicationStatus.Processed)
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:147]: This if is unlikely to be ever true (although possible) as the only statements likely to cause an exception are SaveChanges and Update(comm, true).  It is more likely that SaveChanges fails and if that occurs then the update to the comm to the error state is also unlikely to occur.  Hence the reason to set the flag to processing at the start of the loop.

File: Apps/Database/src/Delegates/DBCommunicationDelegate.cs (hunk 12 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationDelegate.cs:141]: I think in general explicitly list all the statuses that need to be returned/processed is better as what if we introduce a new status in the future that we don't need to return here? Also what about "Pending" communication even we don't have any plan to use this "Pending" status.

File: Apps/Database/src/Delegates/DBCommunicationDelegate.cs (hunk 13 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationDelegate.cs:138]: Do you see any scenarios this can cause issue if we return only the communications based on today and scheduled date? Also for info, schedule date has not been introduced yet.

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs (hunk 19 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:80]: Per MS guidance, method name can be up to 128 length. Method name should describe what it filters and returns. Btw, I will try to come up with a shorter version.
<https://github.com/ktaranov/naming-convention/blob/master/C%23%20Coding%20Standards%20and%20Naming%20Conventions.md>

Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:80]: Updating it to "GetActiveUserProfilesByCommunicationId"

File: Apps/Database/src/Delegates/ICommunicationEmailDelegate.cs (hunk 20 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/ICommunicationEmailDelegate.cs:44]: Updating it to "public List<UserProfile> GetActiveUserProfilesByCommunicationId(Guid communicationId, DateTime? createdOnOrAfter, int maxRows)"

File: Apps/Database/src/Models/Communication.cs (hunk 21 above)
Comment (leobcgov) [reply Apps/Database/src/Models/Communication.cs:82]: this is an old version, not sure why you can still see it there.

File: Apps/Database/src/Delegates/DBCommunicationDelegate.cs (hunk 13 above)
Comment (FuriousLlama) [reply Apps/Database/src/Delegates/DBCommunicationDelegate.cs:138]: Since this is retrieving communications to send I would just name it that. Something like "GetEmailCommunicationsToSend"

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 22 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: Yes, I have tested this as well, It works fine. As @sslaws mentioned, EF core knows to track these objects.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 15 above)
Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:124]: I think the query that is getting the profiles would resolve this successfully without having this additional variable. Unless we want them to be processed by creation date I think we should remove it.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 16 above)
Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:65]: I like fetch size as a name

Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:65]: On that same line, we do need a max retry counter. Otherwise, the loop might not end

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 23 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:81]: Will update it, btw personally I prefer using short name for these in the loop, i.e. foreach(Communication c...) or forearch(Blog b ...). This short name helps the code to be easier to read as It is super hard to read "communication" single object vs. "communications" (list)

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+                this.logger.LogInformation($"Found {communications.Count} communications which need to be processed / re-processed.");
+
+                foreach (Communication communication in communications)
+                {
+#pragma warning disable CA1031 //We want to catch exception.
+                    try
+                    {
+                        DateTime? createdOnOrAfterFilter = null;
+                        string lastProcessedProfileHdid = string.Empty;
+                        List<UserProfile>? usersToSendCommEmails = null;
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:87]: Instead of making it nullable how about initializing it as empty. That way the if below can be simplified

**Comment by leobcgov** [comment]
Code updated. Thanks a lot @FuriousLlama and @sslaws for your great comments and feedbacks. We may need to review some of them on a call today to make sure that I have not missed anything.

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs (hunk 18 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:94]: Leo will look into the approach of using a single LINQ instead of 3 steps here

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+                {
+#pragma warning disable CA1031 //We want to catch exception.
+                    try
+                    {
+                        DateTime? createdOnOrAfterFilter = null;
+                        string lastProcessedProfileHdid = string.Empty;
+                        List<UserProfile> usersToSendCommEmails = new List<UserProfile>();
+                        do
+                        {
+                            if (usersToSendCommEmails != null && usersToSendCommEmails.Count > 0)
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:90]: usersToSendCommEmails cannot be null

**Comment by leobcgov** [comment]
Close this PR to implement a single LINQ query to get the list of User Profiles to create new Communication Email records.
